protocols/luci-proto-bonding/htdocs/luci-static/resources/protocol/bonding.js

```diff
@@ -0,0 +1,442 @@
+			}
+		}
+
+		if (primary_valid == 'false') {
+			return _('You must select a primary interface which is included in selected slave interfaces!');
+		}
+	}
+
+	return true;
+}
+
+return network.registerProtocol('bonding', {
+	getI18n: function() {
+		return _('Link Aggregation (Channel Bonding)');
+	},
+
+	getIfname: function() {
+		return null;
+	},
+
+	getOpkgPackage: function() {
+		return 'bonding';
+	},
+
+	isFloating: function() {
+		return true;
+	},
+
+	isCreateable: function(ifname) {
+		return getSelectableSlaves(ifname).then(L.bind(function(devices) {
+			return devices.length == 0 ?  _('No more slaves available') : null;
+		}, this));
+
+		return _('No more slaves available');
+	},
+
+	isVirtual: function() {
+		return true;
+	},
+
+	getDevices: function() {
+		return null;
+	},
+
+	containsDevice: function(ifname) {
+		return (network.getIfnameOf(ifname) == this.getIfname());
+	},
+
+	renderFormOptions: function(s) {
+		var o;
+
+		o = s.taboption('general', form.Value, 'ipaddr',
+				_('IPv4 address'),
+				_('The local IPv4 address'));
+		o.datatype = 'ip4addr';
+		o.validate = validateEmpty;
+
+		o = s.taboption('general', form.Value, 'netmask',
+				_('IPv4 netmask'),
+				_('The local IPv4 netmask'));
+		o.datatype = 'ip4addr';
+		o.validate = validateEmpty;
+		o.value("255.255.255.0");
+		o.value("255.255.0.0");
+		o.value("255.0.0.0");
+
+		o = s.taboption('advanced', form.MultiValue, 'slaves',
+				_('Slave Interfaces'),
+				_('Specifies which slave interfaces should be attached to this bonding interface'));
+
+		o.load = function(section_id) {
+			return getSelectableSlaves(section_id).then(L.bind(function(devices) {
+				for (var i = 0; i < devices.length; i++) {
+					this.value(devices[i], devices[i]);
+				}
+
+				if (devices.length == 0) {
+					this.placeholder = _('No more slaves available, can not save interface');
+					this.value('', '');
+					return '';
+				}
+
+				var if_slaves = uci.get('network', section_id, 'slaves');
+
+				if (if_slaves != null) {
+					return if_slaves;
+				}
+
+				return;
+			}, this));
+		};
+		o.cfgvalue = function(section_id) {
+			return uci.get('network', section_id, 'slaves');
+		};
+		o.validate = validateEmpty;
+		
+		o = s.taboption('advanced', form.ListValue, 'bonding_policy',
+				_('Bonding Policy'),
+				_('Specifies the mode to be used for this bonding interface'));
+		o.default = 'balance-rr';
+		o.value('balance-rr', _('Round-Robin policy (balance-rr, 0)'));
+		o.value('active-backup', _('Active-Backup policy (active-backup, 1)'));
+		o.value('balance-xor', _('XOR policy (balance-xor, 2)'));
+		o.value('broadcast', _('Broadcast policy (broadcast, 3)'));
+		o.value('802.3ad', _('IEEE 802.3ad Dynamic link aggregation (802.3ad, 4)'));
+		o.value('balance-tlb', _('Adaptive transmit load balancing (balance-tlb, 5)'));
+		o.value('balance-alb', _('Adaptive load balancing (balance-alb, 6)'));
+
+		o = s.taboption('advanced', form.ListValue, 'primary',
```

Comment:
Mabye consider using `luci.tools.widgets.DeviceSelect` here and use its `.filter = function(...) { ... }` callback to only present slave candidates.

Reply:
I think I have a misunderstanding here.
Exept for using the widgets instead the ListValue I see no more advantages.
If I change the slave candidates during the page is shown this will not update the primaries dependencies, isn't it? The primaries list will only be updated on loading the page. Right?
So, I will still need the validate function.
Is there a solution to dynamically update the primaries list?

Reply:
You can update the primaries from the `validate()` callback of another function using something like this technique: https://github.com/openwrt/luci/blob/3c4bc228a1f7b5731cf464f3f407c9ed9ace3cd2/applications/luci-app-firewall/htdocs/luci-static/resources/tools/firewall.js#L420

Reply:
I made it following your suggestion.